### PR TITLE
Add Rust int param types

### PR DESCRIPTION
**Pull request by adamrk**
Adds module param types for `i8`, `i64`, `isize`, and `usize`. This is done using a trait `ModuleParam` which will be expanded to work for custom param types as well. Addresses part of https://github.com/Rust-for-Linux/linux/issues/11.

### CODE REVIEW

File: rust/kernel/module_param.rs
Comment (ojeda) [inline rust/kernel/module_param.rs]: If we are doing parsing in our side for some of the types, I am wondering if we should do it for all our types, i.e. just always use `charp` or our own custom `kernel_param_ops` like here and then the trait does the conversion to whatever the target type is (be it primitives, custom types, etc.). Thoughts?

I can think of two downsides:

File: rust/Makefile
Comment (ojeda) [inline rust/Makefile]: Which symbol triggered this? The generated `$ops`?

Comment (adamrk) [reply rust/Makefile]: Yes, exactly.

File: rust/kernel/module_param.rs
Comment (adamrk) [reply rust/kernel/module_param.rs]: The downsides?

I think if we use `charp` for everything then we'll be doing allocations for types that don't need it and we'd have to add some extra logic if we don't want to re-parse every time the param is read by the module. For those reasons I think using our own `ops` would be better.

Seems like using our own ops for everything would be fine, but we should make sure that we can at least parse everything that's supported for the existing params. For bool we'd need to accept 'yYnN01' and for the integer types we'd need to add support for hex and octal (but maybe we should do that anyway for these types?).

Comment (ojeda) [reply rust/kernel/module_param.rs]: > for the integer types we'd need to add support for hex and octal (but maybe we should do that anyway for these types?)

Exactly, I was thinking of supporting hex, octal and also binary for all of them. The C side doesn't currently, because they just pass a particular format string to `scnprintf()` and I guess they didn't want to complicate the parsing for the others, so they added `hexint` independently... But that seems like a workaround, there is no (functional) reason not to support the different bases for all of the primitive types.

Comment (ojeda) [reply rust/kernel/module_param.rs]: Long-term, the C side could call our safe parsing functions instead ;-)

Comment (adamrk) [reply rust/kernel/module_param.rs]: Added parsing of hex/octal/binary and switch to the Rust param ops for all integer types. Using our own param ops for `bool` will require adding support for the `NOARG` flag (which we'll want for custom params anyway) so I can do a separate PR for that and `charp`.

I tried to split the parsing logic into a separate crate that could be used both here and in the `module` proc macro, but it got a bit tricky: The common logic will need to be compiled for the host so that it can be used in the proc macro and also for the target architecture so that it can be used here. Any ideas on the best way to do that?

Also, just to make sure I'm not mixed up, I think the C side supports parsing hex and octal for all the int params (via `kstrtol`) and `hexint` is just different is that it displays as hex via it's `get` function?

File: rust/kernel/module_param.rs
Comment (ojeda) [inline rust/kernel/module_param.rs]:
```suggestion
/// more than 4 KiB (including an additional null terminator).
```

File: rust/kernel/module_param.rs
Comment (ojeda) [reply rust/kernel/module_param.rs]: We can have a crate that is shared for both host and target -- I guess the need will come up sooner or later. I can take a look, although I am not sure if we should add it for the first PR.

For this particular case though, if we end up using `syn`, we wouldn't need it for proc macros, only for the target. Another option: we add it to `core` ;-) There is https://github.com/rust-lang/rfcs/issues/1098 for that.

And you're 100% right, I inverted `get/set`, the format string is for the display side.

File: rust/kernel/module_param.rs
Comment (ojeda) [inline rust/kernel/module_param.rs]: I assume you went for a trait vs. a generic function to make it look like `from_str_radix` for callers, right?

Comment (adamrk) [reply rust/kernel/module_param.rs]: I think we need to make a trait anyway. There's no trait (at least not in `std`) for types that implement `from_str_radix` or `checked_neg` so if we wanted to make a generic function we'd have to create that trait so we could use it as a constraint on the argument for that function. Seems to me that either way would be reasonable.

File: rust/kernel/module_param.rs
Comment (adamrk) [reply rust/kernel/module_param.rs]: On second thought it actually wasn't much of a change to use our own parser for `bool` so just added it here. I don't think we should do a `charp` replacement just now though. I think the inner type would need to be some type of enum so that we can distinguish between the cases where the param needs to freed or not. But then we'd run into the issue of the `default` field in the macro needing to be a literal. So that should probably wait until `syn` too.

File: rust/kernel/module_param.rs
Comment (ojeda) [reply rust/kernel/module_param.rs]: Right, there is the `num` crate out, but it seems like Rust could use a `PrimitiveInteger` trait or similar.

File: rust/kernel/module_param.rs
Comment (ojeda) [inline rust/kernel/module_param.rs]:
```suggestion
    /// `arg == None` indicates that the parameter was passed without an
    /// argument. If `NOARG_ALLOWED` is set to `false` then `arg` is guaranteed
    /// to always be `Some(_)`.
```

File: rust/kernel/module_param.rs
Comment (ojeda) [inline rust/kernel/module_param.rs]: If this is the page size, then we can say `PAGE_SIZE` instead like you do below (since there are different page sizes).

**Review by ojeda — APPROVED** [review]

**Comment by adamrk** [comment]
Rebased onto the latest changes. Does this need a second review?

File: rust/Makefile
Comment (ojeda) [inline rust/Makefile]:
```suggestion
		| grep -E ' (T|R) ' | cut -d ' ' -f 3 | grep -E '^(__rust_|_R)' \
```

File: rust/kernel/buffer.rs
Comment (ojeda) [inline rust/kernel/buffer.rs:5]: Documentation for this file and for `Buffer` (even if `rustdoc` does not require because we don't export it -- shall we, by the way? i.e. do we expect drivers etc. use `Buffer`?).

File: rust/kernel/module_param.rs
Comment (ojeda) [inline rust/kernel/module_param.rs]: Which `to_string()`? If you are referring to a C API, please mention it (ideally, add a link).

File: rust/kernel/module_param.rs
Comment (ojeda) [inline rust/kernel/module_param.rs:9]:
```suggestion
/// Types that can be used for module parameters.
///
```

i.e. the first line/sentence (the "summary") should be separated.

File: rust/kernel/module_param.rs
Comment (ojeda) [inline rust/kernel/module_param.rs:20]: Summary line like `Whether to allow...` or something like that?

File: rust/kernel/module_param.rs
Comment (ojeda) [inline rust/kernel/module_param.rs:30]: Summary line missing, i.e. to answer what the implementation is supposed to fill here?

File: rust/kernel/module_param.rs
Comment (ojeda) [inline rust/kernel/module_param.rs:40]: Summary line missing, ditto for the others.

File: rust/kernel/module_param.rs
Comment (ojeda) [inline rust/kernel/module_param.rs]:
```suggestion
/// Trait for parsing integers.
///
/// Strings begining with `0x`, `0o`, or `0b` are
```

and reflow the comment please.

File: rust/kernel/module_param.rs
Comment (ojeda) [inline rust/kernel/module_param.rs]:
```suggestion
/// leading `+` or `-` is also permitted. Any string parsed by `kstrtol()` or
/// `kstrtoul()` will be successfully parsed.
///
/// [`kstrtol()`]: https://www.kernel.org/doc/html/latest/core-api/kernel-api.html#c.kstrtol
/// [`kstrtoul()`]: https://www.kernel.org/doc/html/latest/core-api/kernel-api.html#c.kstrtoul
```

(the idea long-term is to have `rustdoc` generate these links automatically to the local docs, like intra-doc links, but for the moment let's show how it would look manually).

File: rust/kernel/module_param.rs
Comment (ojeda) [inline rust/kernel/module_param.rs]: Do these show up in the docs? Should they? If we want them there, I guess we should improve the docs.

**Comment by ojeda** [comment]
> Rebased onto the latest changes.

Yeah sorry, we should have merged this before -- anyway this gives a chance to review the docs; the person who writes some code is the best one that can write the documentation anyway :)

File: rust/Makefile
Comment (adamrk) [reply rust/Makefile]: Wouldn't this end up matching symbols that aren't in text or ro data sections if they happen to contain a `T` or `R`?

Comment (ojeda) [reply rust/Makefile]: Note the spaces outside the group; it is an equivalent expression.

Comment (adamrk) [reply rust/Makefile]: yep, missed that.

File: rust/kernel/buffer.rs
Comment (adamrk) [reply rust/kernel/buffer.rs:5]: Yeah I think it makes sense to make it public.

File: rust/kernel/module_param.rs
Comment (adamrk) [reply rust/kernel/module_param.rs]: Added a doc about where the param ops are generated that will be included whenever the macro is used and the ability to add docs where the macro is called.

**Comment by adamrk** [comment]
> Yeah sorry, we should have merged this before -- anyway this gives a chance to review the docs; the person who writes some code is the best one that can write the documentation anyway :)

No worries.

File: rust/kernel/buffer.rs
Comment (ojeda) [inline rust/kernel/buffer.rs]:
```suggestion
//! Struct for writing to a pre-allocated buffer with the [`write!`] macro.
```

Intra-doc links (for `core` and `alloc`, not `std` since we don't use it) should work automatically (or did you try and it was not working?).

File: rust/kernel/lib.rs
Comment (ojeda) [inline rust/kernel/lib.rs:53]:
```suggestion
/// Page size defined in terms of the `PAGE_SHIFT` macro from C.
///
/// [`PAGE_SHIFT`]: ../../../include/asm-generic/page.h
```

File: rust/kernel/module_param.rs
Comment (ojeda) [inline rust/kernel/module_param.rs]:
```suggestion
/// Note that displaying the type in `sysfs` will fail if
/// [`alloc::string::ToString::to_string`] (as implemented through the
/// [`core::fmt::Display`] trait) writes more than [`PAGE_SIZE`]
/// bytes (including an additional null terminator).
```

Ditto: i.e. the Rust intra-doc links are found automatically by rustdoc.

I think `PAGE_SIZE` will also work without `kernel::`, but please double-check.

File: rust/kernel/buffer.rs
Comment (adamrk) [reply rust/kernel/buffer.rs]: Ah, I didn't try it. Was just following the pattern for existing links.

Looks like items in `core` and `alloc` link to `doc.rust-lang.org` even without explicitly setting them. We're not generating our own docs for  `core` and `alloc` - at least I don't see any rules for them in the Makefile and they don't seem to be generated by `make rustdoc`.

File: rust/kernel/module_param.rs
Comment (adamrk) [reply rust/kernel/module_param.rs]: Removed the links `rust-lang.org`, but `PAGE_SIZE` needs to be referenced as `crate::PAGE_SIZE`.

File: rust/kernel/buffer.rs
Comment (ojeda) [reply rust/kernel/buffer.rs]: Yeah, they go to the official docs at the moment, but perhaps we should generate them too. I can look into that. Let me create an issue.

Comment (ojeda) [reply rust/kernel/buffer.rs]: #97

**Comment by ojeda** [comment]
Let's get this in!

Thanks a lot for all the work and enduring the reviews!

**Pull request merged**
